src/object/cli_ec.c

```diff
@@ -33,6 +33,1229 @@
+					true);
+		}
+	}
+
+	if (update) {
+		for (i = 0; i < ec_recx_array->oer_nr; i++) {
+			full_ec_recx = &ec_recx_array->oer_recxs[i];
+			full_recx = &full_ec_recx->oer_recx;
+			ec_parity_recx_add(full_recx, riod->iod_recxs, ridx,
+					   tgt_recx_idxs, oca);
+		}
+		ec_parity_seg_add(ec_recx_array, iod, oca, sorter);
+	}
+
+	obj_ec_seg_pack(sorter, rsgl);
+
+	/* generate the oiod/siod */
+	tgt_nr = update ? obj_ec_tgt_nr(oca) : obj_ec_data_tgt_nr(oca);
+	for (i = 0, idx = 0, last = 0; i < tgt_nr; i++) {
+		/* get each tgt's idx in the compact oiod_siods array */
+		if (tgt_recx_nrs[i] != 0)
+			tidx[i] = idx++;
+		else
+			tidx[i] = -1;
+		for (j = last; j < tgt_recx_idxs[i] + tgt_recx_nrs[i]; j++) {
+			if (riod->iod_recxs[j].rx_nr != 0)
+				continue;
+			/* being merged so left empty space */
+			D_ASSERT(j != tgt_recx_idxs[i]);
+			D_ASSERT(j < riod->iod_nr);
+			for (k = j; k < tgt_recx_nrs[i] + tgt_recx_idxs[i]; k++)
+				D_ASSERT(riod->iod_recxs[k].rx_nr == 0);
+			empty_nr = tgt_recx_nrs[i] + tgt_recx_idxs[i] - j;
+			for (k = j; k < riod->iod_nr - empty_nr; k++)
+				riod->iod_recxs[k] =
+					riod->iod_recxs[k + empty_nr];
+			for (k = riod->iod_nr - empty_nr;
+			     k < riod->iod_nr; k++) {
+				riod->iod_recxs[k].rx_idx = 0;
+				riod->iod_recxs[k].rx_nr = 0;
+			}
+			tgt_recx_nrs[i] -= empty_nr;
+			for (k = i + 1; k < tgt_nr; k++)
+				tgt_recx_idxs[k] -= empty_nr;
+			riod->iod_nr -= empty_nr;
+			break;
+		}
+		last = tgt_recx_idxs[i] + tgt_recx_nrs[i];
+	}
+	oiod->oiod_nr = idx;
+	for (i = 0, rec_nr = 0, last = 0; i < tgt_nr; i++) {
+		if (tgt_recx_nrs[i] == 0)
+			continue;
+		siod = &oiod->oiod_siods[tidx[i]];
+		siod->siod_tgt_idx = i;
+		siod->siod_idx = tgt_recx_idxs[i];
+		siod->siod_nr = tgt_recx_nrs[i];
+		siod->siod_off = rec_nr * iod_size;
+		for (idx = last; idx < tgt_recx_idxs[i] + tgt_recx_nrs[i];
+		     idx++) {
+			rec_nr += riod->iod_recxs[idx].rx_nr;
+		}
+		last = tgt_recx_idxs[i] + tgt_recx_nrs[i];
+	}
+
+#if EC_DEBUG
+	obj_reasb_req_dump(reasb_req, sgl, oca, stripe_rec_nr, iod_idx);
+#endif
+
+	if (iovs != NULL && iovs != iov_inline)
+		D_FREE(iovs);
+	return rc;
+}
+
+int
+obj_ec_req_reasb(daos_obj_rw_t *args, daos_obj_id_t oid,
+		 struct daos_oclass_attr *oca, struct obj_reasb_req *reasb_req,
+		 bool update)
+{
+	daos_iod_t		*iods;
+	d_sg_list_t		*sgls;
+	uint32_t		 iod_nr = args->nr;
+	int			 i, rc = 0;
+
+	iods = args->iods;
+	sgls = args->sgls;
+	for (i = 0; i < iod_nr; i++) {
+		rc = obj_ec_recx_scan(&iods[i], &sgls[i], oca, reasb_req, i,
+				      update);
+		if (rc) {
+			D_ERROR(DF_OID" obj_ec_recx_scan failed %d.\n",
+				DP_OID(oid), rc);
+			goto out;
+		}
+
+		rc = obj_ec_recx_encode(oid, &iods[i], &sgls[i], oca,
+					&reasb_req->orr_recxs[i]);
+		if (rc) {
+			D_ERROR(DF_OID" obj_ec_recx_encode failed %d.\n",
+				DP_OID(oid), rc);
+			goto out;
+		}
+
+		rc = obj_ec_recx_reasb(&iods[i], &sgls[i], oca, reasb_req, i,
+				       update);
+		if (rc) {
```

Comment:
Maybe switch the recx_reasb() with encode(), then after scan and rehabs(), the data will be ready to send to the server.  There are a few advantages for this

1. Once we are ready for client to forward the requests, instead of by leader, then the data can be forward to the server directly, instead of waiting for the EC encoding.   Plus that we may use those firmware to compute EC asynchronously.
2. Separate the split and EC, then later the array object can share the same code for split.
3. If you do scan & resab together,  it does not need to track the full stripe status, which is only need for EC, then it will simplify the code.